SoundSource Plugin API: Version 8

uklotzde opened the pull request:
Fixes https://bugs.launchpad.net/mixxx/+bug/1452651

File names for SoundSource plugins were truncated at the first occurrence of '#' due to an implicit conversion from QString to QUrl.

uklotzde committed:
Log warning about unreadable tags in audio files

uklotzde committed:
Change parameter type of getSoundSource() from QString to QUrl

uklotzde committed:
Merge branch 'master' into SoundSourcePluginUrlFileNameFix

uklotzde committed:
Add an assertion to detect invalid conversions QString <-> QUrl

uklotzde committed:
New streamlined SoundSource plugin API version 8

A simple C++ API that reduces the chance of memory leaks that were
imminent in the previous C/C++ version. Error handling and logging
in SoundSourceProxy while loading plugins has been improved.

uklotzde commented:
I've bumped the SoundSource plugin API version from 7 to 8, re-defined the public library functions, and did some cleanup.

uklotzde committed:
Detach dynamic library management from SoundSourceProxy

uklotzde committed:
Introduce SoundSourceProvider interface

SoundSourceProvider is a generic interface that will implemented both by
plugin libraries as well as built-in sound sources. This will simplify
the selection of an appropriate SoundSource in SoundSourceProxy and is
useful for future extensions like selecting one out of multiple providers
for each file type.

uklotzde committed:
Add and register a SoundSourceProvider for each SoundSource

uklotzde committed:
Fix FFmpeg build

uklotzde committed:
Give SoundSource providers a name

uklotzde committed:
Delete obsolete #include directive

uklotzde committed:
Wrap static members from SoundSourceProxy into class

uklotzde committed:
Fix static initialization of SoundSource providers in tests

uklotzde committed:
Fix SoundSource plugin warning

uklotzde committed:
Log detailed debug info about SoundSource providers

uklotzde committed:
Move final debug logging out of loop

uklotzde committed:
Delete obsolete member from SoundSourceProxy

uklotzde committed:
Avoid unnecessary locking in SoundSourceProxy

uklotzde committed:
Eliminate duplicate initialization in tests

uklotzde committed:
Add debug logging for seek test

uklotzde commented:
What has begun as a simple bug fix is now a refactoring of the SoundSource plugin framework ;)

uklotzde committed:
Simplify SoundSource plugin registration

uklotzde committed:
Delete obsolete member functions and code cleanup

uklotzde committed:
Use plugin file name filter only for Apple OS X

At least this is how it works in 1.12.

daschuer commented on src/sources/urlresource.h:
```diff
@@ -19,11 +19,16 @@ class UrlResource {
+    }
+
     inline QString getLocalFileName() const {
+        DEBUG_ASSERT(isLocalFile());
```
What happens on windows with network shares? 
There is a pending bug for SMB locations 
https://bugs.launchpad.net/mixxx/+bug/918222
Will this fail here?

daschuer commented on src/sources/soundsourcepluginlibrary.cpp:
```diff
@@ -0,0 +1,82 @@
+/*static*/ QMap<QString, SoundSourcePluginLibraryPointer> SoundSourcePluginLibrary::s_loadedPluginLibraries;
+
+/*static*/ SoundSourcePluginLibraryPointer SoundSourcePluginLibrary::load(
+        const QString& fileName) {
```
libFilePath

daschuer commented on src/sources/soundsourcemp3.cpp:
```diff
@@ -700,4 +694,14 @@ SINT SoundSourceMp3::readSampleFrames(
 }
 
+QString SoundSourceProviderMp3::getName() const {
+    return "libmad";
```
This is official called "MAD: MPEG Audio Decoder" or just "MAD"

daschuer commented on src/soundsourceproxy.cpp:
```diff
@@ -47,17 +20,24 @@
+#ifdef __APPLE__
+const QStringList SOUND_SOURCE_PLUGIN_FILENAME_PATTERN("libsoundsource*");
+#else
+// NOTE(uklotzde): Why no filtering for Linux/Windows?
```
On Linux the filter should work as well or do I miss something?  
In windows the lib postfix might be missing.

daschuer commented:
Nice work! Thank you very much. 
Linux version works like a charm. 

I think we need a Windows and Mac test before merge.
Any volunteers? 

Does the build server produce plug-Ins? For win and Mac?

uklotzde committed:
Change parameter name: fileName -> libFilePath

uklotzde committed:
Change getter name: getFileName() -> getFilePath()

The corresponding Qt function name is actually misleading.

uklotzde committed:
Change the name of the MP3 SoundSource provider

uklotzde committed:
Add a TODO note about network shares

According to the documentation of QUrl::isLocalFile()/toLocalFile()
network shares are handled in the same way as before, i.e. they are
considered to be local files with all consequences.

uklotzde committed:
Filtering of plugin file names on UNIX systems (not only OS X)

uklotzde committed:
Merge branch 'master' into SoundSourcePluginUrlFileNameFix

uklotzde committed:
Streamline the MetadataSource interface

Usually both metadata and cover art should be parsed at once. Optional
parameters are used if only one or the other is needed. A single
virtual function covers all those use cases.

uklotzde committed:
Delete obsolete member function from TrackDAO

uklotzde committed:
Add checks for supported file types to SoundSourceProxy

uklotzde committed:
Remove obsolete local variables

uklotzde committed:
Consistent naming: "file type" instead of "file extension"

uklotzde committed:
Delete obsolete functions from SoundSourceFFmpeg

daschuer commented:
@default-kramer: If you have some time, would you mind to jump in here and test the windows part?  
Thank you.

daschuer commented on src/sources/soundsourcemodplug.cpp:
```diff
@@ -86,22 +73,16 @@ Result SoundSourceModPlug::parseTrackMetadata(
 
-QImage SoundSourceModPlug::parseCoverArt() const {
-    // The modplug library currently does not support reading cover-art from
-    // modplug files -- kain88 (Oct 2014)
```
This comment is lost.

daschuer commented on src/trackinfoobject.cpp:
```diff
@@ -255,7 +255,8 @@ void TrackInfoObject::parse(bool parseCoverArt) {
         Mixxx::TrackMetadata trackMetadata;
-        if (proxy.parseTrackMetadata(&trackMetadata) == OK) {
+        QImage* pCoverArt = parseCoverArt ? &m_coverArt.image : NULL;
+        if (proxy.parseTrackMetadataAndCoverArt(&trackMetadata, pCoverArt) == OK) {
```
This took me a while to understand. 
Please do something to help.  
Add a comment or replace this with if/else + two calls for parseTrackMetadataAndCoverArt. 
.. or could we just add two wrappers parseTrackMetadata() and parseCoverArt().

daschuer commented on src/library/autodj/autodjfeature.cpp:
```diff
@@ -170,9 +170,8 @@ bool AutoDJFeature::dropAccept(QList<QUrl> urls, QObject* pSource) {
-    return SoundSourceProxy::isFilenameSupported(file.fileName()) ||
-            Parser::isPlaylistFilenameSupported(file.fileName());
+    return SoundSourceProxy::isUrlSupported(url) ||
+            Parser::isPlaylistFilenameSupported(url.toLocalFile());
```
What is the reason for the double check? Can one be true and the other false?

daschuer replied:
Ah never mid. The second check is  for playlists.

daschuer commented on src/dlgpreflibrary.cpp:
```diff
@@ -59,7 +59,7 @@ DlgPrefLibrary::DlgPrefLibrary(QWidget * parent,
     // plugins are loaded in src/main.cpp way early in boot so this is safe
     // here, doesn't need done at every slotUpdate
-    QStringList plugins(SoundSourceProxy::supportedFileExtensionsByPlugins());
+    QStringList plugins(SoundSourceProxy::getSupportedFileTypesByPlugins());
```
I liked the old version more. We just check for the file extension, not the mime type or something from libmagic. 
Inside Mixxx we use inconsistent getter style some have get prefix some not. Qt does not use the get prefix ,,, just a matter of taste.

uklotzde replied:
There already was an inconsistency between "type" and "extension" before I started to refactor the SoundSource API. I have chosen the more generic term "type" instead of "extension", because in the future we might improve the deduction of the file type. For the time being simply the file extension is used as the type of the file.

uklotzde committed:
Fix indentation

uklotzde committed:
Fix and comment parsing of cover art

uklotzde committed:
Restore deleted comment

uklotzde committed:
Improve comment in MetadataSource interface

uklotzde committed:
Add and improve functions for backward compatibility

daschuer commented:
Thank you @uklotzde  LGTM 
Do you have a clue how the test coverage is on Linux and Windows? 

@rryan: unfortunately building of master has stopped. Can you have a look? 
Is it OK to merge this and test the master builds?

uklotzde commented:
The SoundSource plugins are loaded and tested by the SoundSourceProxyTest cases. But there are no separate tests for the plugin framework yet.

uklotzde commented:
@daschuer: Your remarks about file "type" and "extension" were correct. The providers actually deal with file extensions, because we are constructing file name patterns from those. I will change the naming and document the decision to simply use the file extension as the file type.

uklotzde committed:
Correctly distinguish between "file extension" and "SoundSource type"

uklotzde commented:
I will add an overridable priority to SoundSourceProviders to support registering of multiple providers with different priorities for the same file extension. See also: https://bugs.launchpad.net/mixxx/+bug/1461819

With this extension SoundSourceFFmpeg can easily be added as a fallback for many file extension instead of just replacing all other SoundSources.

daschuer commented:
Since this is already pretty mature, can you add additional changes in a separate PR on top of this? 

@rryan: I would like to merge this, if you do not have any objections. 
We may fix Mac and Windows issues directly in master.

uklotzde commented:
@daschuer: Just give me a minute, I will just modify the SoundSourceProvider interface and push the changes (local build is in progress). The actual of the implementation can be done in a separate branch.

uklotzde commented:
Done, see my last commit. Ready for merging after the CI build has finished successfully.

uklotzde commented:
Ok, now it's done ;) That would not have happened with C++11 and the new "override" keyword!

daschuer commented:
I am in doubt if that priority concept will carry us for long. 
The plug-in itself cannot decide over its own priority unless it knows the other plug-Ins.

Mixxx can do it instead with a simple set of rule: 
"Use a non FFMPEG  plug-In if it fails use FFMPEG"
This is only true on Linux, isn't it? And How a abound the situation if the FFMPEG stack gives the best results for a single file type ...
This seams also not that is not the best solution. 
But it will work, because today, Mixxx actually knows all possible SoundSource plugIns and it would give use the feedom to decide what is best per OS and file type. 

Later, this could be a improved by something like this: 
The API can provide a feature indicator, that tells Mixxx how well a certain file type is supported. 
Mixxx can than just choose the best soundsource for each type.   

For now I would prefer to merge this branch without the additional priority API. It should be no problem to switch the API version again in the next PR if required.

What do you think about this?

uklotzde commented:
@daschuer: I've reset the branch to the previous state without the priority extension. Adding the priority later will require to bump the plugin version from 8 to 9, but this is ok. We will not run out of version numbers soon ;)

The priority is only a "hint" to provide a reasonable default behavior without requiring any extra configuration by the user or some sort of GUI.

uklotzde committed:
Initialize FFmpeg before registering providers and loading plugins

rryan commented on src/library/coverartutils.h:
```diff
@@ -36,9 +36,7 @@ class CoverArtUtils {
-        if (trackLocation.isEmpty()) {
-            return QImage();
-        }
+        DEBUG_ASSERT(!trackLocation.isEmpty());
```
Generally we should prefer DEBUG_ASSERT_AND_HANDLE.

```DEBUG_ASSERT_AND_HANDLE(...) {
  return QImage();
}```

uklotzde replied:
Empty file names and non-existent files are handled gracefully by SoundSourceProxy, we don't need to handle these errors here again. Instead of handling it twice I would prefer to remove the DEBUG_ASSERT entirely.

rryan commented on src/test/mixxxtest.cpp:
```diff
@@ -9,60 +9,43 @@ extern "C" {
-}
 
-MixxxTest::MixxxTest() {
-    // Create QApplication as a singleton. This prevents issues with creating
```
Please keep the comment.

rryan replied:
ping :)

uklotzde committed:
Delete redundant DEBUG_ASSERT

Empty file names are handled gracefully by SoundSourceProxy.

rryan commented on src/sources/soundsourceprovider.h:
```diff
@@ -0,0 +1,26 @@
+
+#include "sources/soundsource.h"
+
+#include <QStringList>
```
Missing include for QUrl, QSharedPointer, QString

rryan commented:
Thanks @uklotzde ! 

When Albert and I worked on the original plugin API we took care not to pass heap allocated objects across the DLL / binary boundary and free them on the other side. On Windows, this will cause immediate segfaults if the DLL and the binary are linked to different runtimes. You can see an example of this style with the old ```supportedFileExtensions/freeFileExtensions``` methods. 

I notice that the 1.12 branch (pre-SS-refactor) commits a huge sin here by allocating the heap-allocated SoundSource itself in the plugin and freeing it in Mixxx so I think we've gotten sloppy over the years in this regard. 

I'm worried about the use of QSharedPointer/QStrings in this version of the API. When the DLL allocates a QSharedPointer/QString and passes it to the main Mixxx binary, I believe the code executing to handle that object is the Mixxx binary's copy of QString/QSharedPointer code. Correct me if I'm wrong!

So when the QSharedPointer<SoundSourceProvider> destructor runs, the Mixxx binary QSharedPointer destructor is what is actually running, which will free the SoundSourceProvider using the Mixxx heap, not the DLL heap. 

Have you given this any thought?

uklotzde commented:
@rryan 
Well, I know that special care has to be taken with DLLs and memory allocation. But my development experience on Windows lies more than 10 years behind in the past ;)

As you mentioned I've seen that many resources are allocated on one side and freed on the other: The SoundSource itself, QString with metadata, QImage with cover art, .... Only doing it right for the file extensions didn't seem to make any sense to me?! So I assumed that this doesn't matter here.

uklotzde committed:
Restore comment about QApplication in tests

uklotzde committed:
Reorder and add missing #include directives

uklotzde committed:
Static allocation of SoundSourceProvider in plugins

uklotzde committed:
Symmetrically allocate and delete SoundSource from within plugins

uklotzde commented:
The C++ operators new and delete are now called symmetrically from within the external plugin libraries and should use the same C/C++ runtime. Unfortunately I'm not able to test if this works as expected for Windows DLLs.

uklotzde committed:
Merge branch 'master' into SoundSourcePluginUrlFileNameFix

uklotzde committed:
Use C+11 override keyword

uklotzde committed:
Add comments about singleton/shared SoundSourceProvider instances

uklotzde commented:
@rryan 
All invocations of operator new are now complemented by the corresponding invocation of operator delete within the same DLL. Allocations within Qt shouldn't matter here if Qt is doing it correctly in their DLL.

daschuer commented on plugins/soundsourcem4a/soundsourcem4a.cpp:
```diff
@@ -430,27 +423,50 @@ SINT SoundSourceM4A::readSampleFrames(
+
+// Singleton: SoundSourceProviderM4A is stateless
+// and a single instance can safely be shared
+Mixxx::SoundSourceProviderM4A SOUNDSOURCE_PROVIDER;
```
static variables should be camel case and prefixed with s

daschuer replied:
In general, we should avoid static variables in libraries. 
Can't we make it a static local of Mixxx_SoundSourcePluginAPI_getSoundSourceProvider

uklotzde replied:
@daschuer Ok, using a static local variable is a viable option

uklotzde committed:
Unify export of SoundSources from plugin libraries

daschuer commented:
> Allocations within Qt shouldn't matter here if Qt is doing it correctly in their DLL.

I want to try to understand the issue. Here some thoughts, please correct me If I am wrong.  

I wonder how Qt can do this with implicit sharing lets say in QString. 
I think this will be linked by the Application and by the DLL to the QT DLL with their matching, but different memory model, finally to different implementations to malloc. 
If now a QString is transfered from the APP to the dll, the QString content is allocated in the senders 
heap model but might be deleted in the receivers model -> crash!

Is that correct? 

If yes, We can use Qt containers only, if the linked used QT dlls are compiled with the same memory model. This means we cannot switch the memory model if we (only hypothetically) split out the plug-in builds. Since we can't do this anyway, we do not need to worry about a different memory models. 

So we have finally no issues here, Right?

uklotzde commented:
@daschuer Both app and plugin must be linked against the same shared Qt DLL. Then alll internal invocations of operator new/delete in Qt classes use the same C++ runtime and match. What happens in this library stays in this library ;)

We only need to take care of managed pointers. They have to be fed with the appropriate deleter upon construction. This ensures that operator delete from the plugin is called instead of that from the app. This approach should work even when the plugins are compiled separately.

uklotzde committed:
Prefer local static variables for shared data in plugins

daschuer commented:
Thanks for clarifying!

rryan commented:
I'm considering the case where some external party releases a soundsource plugin DLL. I had been assuming they were statically linking their own version of Qt -- but I guess that's prone to causing other problems anyway so we may require that they dynamically link to Qt.

If we accept that they have to dynamically link to Qt (so they are using our Qt DLLs once we load their DLL) -- then I agree the heap used for shared Qt objects will be the one we chose when we compiled Qt.

rryan commented:
Thanks for the changes @uklotzde!
